index.js

```diff
@@ -12,23 +12,23 @@ var validate = require('./lib/validate');
   * @param {Function}        callback Returns error or project data
   */
 module.exports = function (input, isSprite, callback) {
-    // First unpack the input (need this outside of the async waterfall so that
-    // unpackedProject can be refered to again)
-    unpack(input, isSprite, function (err, unpackedProject) {
-        if (err) return callback(err);
-
-        async.waterfall([
-            function (cb) {
-                parse(unpackedProject[0], cb);
-            },
-            // TODO is there a better way to pass this arg
-            // than partially applying this funciton?
-            validate.bind(null, isSprite)
-        ], function (error, validatedInput) {
-            // One more callback wrapper so that we can re-package everything
-            // with the possible zip returned from unpack
-            if (error) return callback(error);
-            callback(null, [validatedInput, unpackedProject[1]]);
-        });
-    });
+    // Unpack the input and further transform the json portion by parsing and
+    // validating it.
+    var toUnpack = unpack(input, isSprite);
+    /* global Promise */
+    Promise.all([
+        toUnpack
+            .then(function (unpackedProject) {
+                return unpackedProject[0];
+            })
+            .then(parse)
+            // TODO is there a better way to pass this arg than partially
+            // applying this function?
+            .then(validate.bind(null, isSprite)),
+        toUnpack
+            .then(function (unpackedProject) {
```

Comment:
This new structure seems a bit awkward now, maybe we can reorganize this code so that we have nested `.then`s? E.g. the other functions e.g. `parse`, `validate`, etc. live inside the `.then` for the `unpack` and then the last bit has access to the `unpackedProject` and can call the callbacks...

Reply:
??

I think what you described is nesting thens.

```js
Promise.all([
    toUnpack
        .then(function (unpacked) {
            return parse(unpacked[0])
                .then(validate.bind(null, isSprite));
        }),
    toUnpack
        .then(function (unpacked) {
            return unpacked[1];
        })
])
    .then(callback.bind(null, null), callback);
```

How about:

```js
unpack(input, isSprite)
    .then(function (unpacked) {
        return Promise.all([parse(unpacked[0]), unpacked[1]]);
    })
    .then(function (parsed) {
        return Promise.all([validate(isSprite, parsed[0]), parsed[1]]);
    })
    .then(callback.bind(null, null), callback);
```

I normally try to write promise functions that do not need external contexts. It probably doesn't matter here since it needs so few lines, but its a big help with large promise chain functions.

On that point what about:

```js
unpack(input, isSprite)
    .then(function (unpackedProject) {
        return parse(unpackedProject[0])
            .then(validate.bind(null, isSprite))
            .then(function (validatedProject) {
                return [validatedProject, unpackedProject[1]];
            });
    })
    .then(callback.bind(null, null), callback);
```

Reply:
@mzgoddard, that last one is pretty much what I was trying to suggest. Sorry I should have been more clear.

The `callback.bind(null, null)` is still really weird to me though, especially as the `resolve` function for the `then`